libraries/rush-lib/src/logic/operations/OperationExecutionRecord.ts

```diff
@@ -320,18 +342,33 @@ export class OperationExecutionRecord implements IOperationRunnerContext, IOpera
             ? OperationStatus.NoOp
             : OperationStatus.Skipped;
       }
-      // Delegate global state reporting
-      await onResult(this);
     } catch (error) {
       this.status = OperationStatus.Failure;
       this.error = error;
-      // Delegate global state reporting
-      await onResult(this);
     } finally {
+      // We may need to clean up and finalize resources (_operationMetadataManager for example needs to be saved by CacheableOperationPlugin)
+      await beforeResult(this);
+      if (this.isTerminal) {
+        this.stopwatch.stop();
+        if (
+          !this.executedOnThisAgent &&
+          this.nonCachedDurationMs &&
+          this.status !== OperationStatus.FromCache
+        ) {
+          const { startTime } = this.stopwatch;
+          if (startTime) {
+            this._stopwatch = Stopwatch.fromState({
+              // use endtime as it's the more accurate version of when this operation was marked as complete.
+              startTime,
+              endTime: startTime + this.nonCachedDurationMs
+            });
+          }
+        }
```

Comment:
I'd feel more comfortable just removing `readonly` from `this.stopwatch` and letting the cobuild logic do this externally. It'd also be more correct to compute the timing by setting `startTime = Math.max(0, stopwatch.endTime - this.nonCachedDurationMs)`.

That said, can we instead update the timeline summarizer to report both cached and uncached durations in its summary? I don't mind the stopwatch tracking the duration from when the operation was first considered until the time it completed; that's a useful metric.